### PR TITLE
Adding support for .yarnrc.yml files.

### PR DESCRIPTION
**Pull request by dannyvv**
This change factors commonalities between .npmrc and .yarnrc files into a common base class `FileProvider`.
This allowed for the addition of .yarnrc.yml files without duplicating code.

This caused some of the logic to move around a bit. The biggest change for the npmrc is a more genereralized approach to reading, writing and updating the file without regular expressions and multiple passes.

Fixes: #4

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,5 @@ node_modules
 .store
 dist 
 lib
-.npmrc
\ No newline at end of file
+.npmrc
+.vscode
```
Comment (dannyvv) [inline .gitignore:6]: @jcreamer898 : Let me know if you prefer all of `.vscode` or just `.vscode/launch.json`

Comment (jcreamer898) [reply .gitignore:6]: Probably just `launch.json`.

Comment (dannyvv) [reply .gitignore:6]:
```suggestion
.vscode/launch.json
```

File: packages/ado-npm-auth/src/fileProvider.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  public async validateAllUsedFeeds(): Promise<ValidatedFeed[]> {
+    await this.prepUserFile();
+
+    const result: ValidatedFeed[] = [];
+
+    const workspaceRegistries = await this.getWorkspaceRegistries();
+    const userFeeds = await this.getUserFeeds();
+
+    // check each feed for validity
+    for (const registry of workspaceRegistries) {
```
Comment (jcreamer898) [inline packages/ado-npm-auth/src/fileProvider.ts:64]: I don't know what the perf of having multiple registries and doing all these calls in a loop might be, but if we test it and find out it's a little slow, we could do all of this work in a `Promise.all`

**Review by jcreamer898 — CHANGES_REQUESTED** [review]
A few minor things, but overall, WOW! Thanks a ton for this!

**Comment by jcreamer898** [comment]
You also might wanna just go ahead and run `yarn prettier --write src/**/*.ts` as I saw some formatting and missing `;`'s etc.

File: packages/ado-npm-auth/src/cli.ts [hunk 3]
```diff
@@ -2,32 +2,87 @@ import { isSupportedPlatformAndArchitecture } from "./azureauth/is-supported-pla
...
+import { YarnRcFileProvider } from "./yarnrc/yarnrcFileProvider.js";
 
-export const run = async (): Promise<null | boolean> => {
-  const doValidCheck = !process.argv.includes("--skip-check");
-  const skipAuth = process.argv.includes("--skip-auth");
+const fileProviders = [new NpmrcFileProvider(), new YarnRcFileProvider()]
+
+export const run = async (args: Args): Promise<null | boolean> => {
+  
+  let validatedFeeds: ValidatedFeed[] = [];
```
[the remark this answers is not in the document]
Comment (dannyvv) [reply packages/ado-npm-auth/src/cli.ts:16]:
```suggestion

  const validatedFeeds: ValidatedFeed[] = [];
```

File: packages/ado-npm-auth/src/fileProvider.ts (hunk 2 above)
Comment (dannyvv) [reply packages/ado-npm-auth/src/fileProvider.ts:64]: The existing code worked similar to this and performed it linearly...
The first part of the logic checks for file existance to see if a fileProvider is applicable, so it should be fast and only affect repo's with multiple configs...

This particular is just a quick http check. restoring packages makes 1000's of these. I have no concern here.

The slowest part in the whole process is requestiong a new PAT (i.e. running the executable) and that is reused as much as possible by deduping ADO orgs among multiple feeds and fileproviders.
Even though it only would affect repo's with feeds going to multiple ADO orgs, this part could perhaps be done in parallel. On apple it pops up something in the browser though, not sure if that actually works with 2 instances running in parralel, but we can try if we get complaints....

**Comment by dannyvv** [comment]
> You also might wanna just go ahead and run `yarn prettier --write src/**/*.ts` as I saw some formatting and missing `;`'s etc.

I'll add 'lint' support and run prettier seperately.

**Review by jcreamer898 — APPROVED** [review]
👏

**Pull request merged**